### DISCUSSION TITLE
Nested class function

### DISCUSSION THREAD

**Discussion by rickypid** [typegoose/typegoose#460] [Q&A] [answered]
## Versions

- System: windows 
- NodeJS: 12.6.1
- Typescript: 4.1.3

- Compiler gulp-typescript
- Typegoose(NPM): 7.4.5 
- Typegoose(GIT): commithash
- mongoose: 5.11.8 
- mongodb: 0.0.0

---

![image](https://user-images.githubusercontent.com/19270983/103110406-a7657a80-4647-11eb-9025-207152c872a8.png)

![image](https://user-images.githubusercontent.com/19270983/103110819-c6640c80-4647-11eb-8986-6bb4476db4dd.png)

I would need to access a function of the nested class. it's possible? am I using it correctly?

**Comment by smolinari** [comment, marked as answer]
@rickypid - First off, refs are only for data reference (thus the name refs), i.e. creating relationships between collections. Secondly, `this` only refers to the current class/ function instance, which test isn't a part of (and won't be via refs)  and why you are getting the error. What you might be looking for is simple inheritance. So, class `Main` could be extending class `Nested`. Though, the method wouldn't be `this.nested.test()` , it would be rather just `this.test()`.

Also, entities really shouldn't be holding any kind of business logic If that is the intention of `test()`. They should be considered just blueprints for something like "containers of data", which are injected and transformed into models, which in turn produce documents (via Mongoose). You can add instance methods, but they should be simple and deal mainly with the data of the instance.

Lastly, I'd suggest if you aren't sure of how something works, you should ask questions in the Discord chat. Githup should be reserved for bugs and improvement suggestions. 

https://discord.com/channels/612400042900193311/612406912465502224

Scott

**Comment by rickypid** [comment]
thank you very much for the very fast reply. So if I just wanted to have a nested class would that be enough?
![image](https://user-images.githubusercontent.com/19270983/103131715-2091bb80-46a2-11eb-84ef-3d34b81112b1.png)

**Comment by hasezoey** [comment]
@rickypid yes, this is an *subdocument* and not an *reference* anymore, if that is what you want

---

@smolinari is right, but i want to add things, firstly there is also Github Discussions now (where i will move this issue to) and secondly:
- it looks like you are using an unsupported version of mongoose, highest supported is 5.10.18
- if the functions in the classes are meant to be on the model (static) or on the document (instance), then you need to overwrite the `this` value, [as can be seen here](https://typegoose.github.io/typegoose/docs/guides/quick-start-guide#extra-examples)
- if you are trying to call / get something from references, then you first need to check if the reference is populated or not, for that there are these functions: [`isDocument`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document) and [`isRefType`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-ref-type)
